homeassistant/components/alarm_control_panel/manual.py

```diff
@@ -0,0 +1,142 @@
+"""
+homeassistant.components.alarm_control_panel.manual
+
+Configuration:
+
+alarm_control_panel:
+  platform: manual
+  name: "HA Alarm"
+  code: "mySecretCode"
+  pending_time: 60
+  trigger_time: 120
+
+Variables:
+
+name
+*Optional
+The name of the alarm. Default is 'HA Alarm'.
+
+code
+*Optional
+If defined, specifies a code to arm or disarm the alarm in the frontend.
+
+pending_time
+*Optional
+The time in seconds of the pending time before arming the alarm.
+Default is 60 seconds.
+
+trigger_time
+*Optional
+The time in seconds of the trigger time in which the alarm is firing.
+Default is 120 seconds.
+
+"""
+
+import logging
+import time
+import homeassistant.components.alarm_control_panel as alarm
+
+from homeassistant.const import (
+    STATE_ALARM_DISARMED, STATE_ALARM_ARMED_HOME, STATE_ALARM_ARMED_AWAY,
+    STATE_ALARM_PENDING, STATE_ALARM_TRIGGERED)
+
+_LOGGER = logging.getLogger(__name__)
+
+DEPENDENCIES = []
+
+DEFAULT_ALARM_NAME = 'HA Alarm'
+DEFAULT_PENDING_TIME = 60
+DEFAULT_TRIGGER_TIME = 120
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """ Sets up the manual alarm platform. """
+
+    add_devices([ManualAlarm(
+        hass,
+        config.get('name', DEFAULT_ALARM_NAME),
+        config.get('code'),
+        config.get('pending_time', DEFAULT_PENDING_TIME),
+        config.get('trigger_time', DEFAULT_TRIGGER_TIME),
+        )])
+
+
+# pylint: disable=too-many-arguments, too-many-instance-attributes
+class ManualAlarm(alarm.AlarmControlPanel):
```

Comment:
I would prefer this class to be setup based on booleans representing state instead of having methods mutating the internal `_state` attribute with sleeps (which are blocking HA threads).

Non tested code to show an example. Based on having the `state` property be smart and return certain states based on how old a state is.

```python
def __init__(…):
    self._state = False
    self._state_set = None
    self._pending_time = datetime.timedelta(seconds=pending_time)

def alarm_disarm(self, code=None):
    self._state = STATE_ALARM_DISARMED
    self._state_set = dt_util.utcnow()
    self.update_ha_state()

def alarm_arm_home(self, code=None):
    self._state = STATE_ALARM_ARMED_HOME
    self._state_set = dt_util.utcnow()
    track_point_in_time(self.hass, self.update_ha_state, dt_util.utcnow() + self.pending_time)

@property
def state(self):
    if self._state == STATE_ALARM_ARMED_HOME and dt_util.utcnow() - self._state_set < pending_time:
            return STATE_ALARM_PENDING
    return self._state
```